pike/smb2.py

```diff
@@ -1984,6 +2123,75 @@ class FileInformation(Information):
 class FileSystemInformation(Information):
     info_type = SMB2_0_INFO_FILESYSTEM
 
+class FileSecurityInformation(FileInformation):
+    file_information_class = FILE_SECURITY_INFORMATION
+    info_type = SMB2_0_INFO_SECURITY
+
+    def __init__(self, parent=None, end=None):
+        FileInformation.__init__(self, parent)
+        self.revision = 0
+        self.sbz1 = 0
+        self.control = 0
+        self.owner_sid_offset = 0
+        self.group_sid_offset = 0
+        self.offset_sacl = 0
+        self.offset_dacl = 0
+        self.other = ''
+        self.end = end
+        self.owner_sid = None
+        self.group_sid = None
+        self.sacl = ''
+        self.dacl = None
+        self._entries = []
+
+    def append(self, e):
+        self._entries.append(e)
+
+    def _decode(self, cur):
+        self.revision = cur.decode_uint8le()
+        self.sbz1 = cur.decode_uint8le()
+        self.control = cur.decode_uint16le()
+        self.offset_owner = cur.decode_uint32le()
+        self.offset_group = cur.decode_uint32le()
+        self.offset_sacl = cur.decode_uint32le()
+        self.offset_dacl = cur.decode_uint32le()
+        if self.offset_owner != 0:
+            # find next non-zero offset in list
+            self.owner_sid = NT_SID()
+            self.owner_sid.decode(cur)
+        if self.offset_group != 0:
+            self.group_sid = NT_SID()
+            self.group_sid.decode(cur)
+        if self.offset_sacl != 0:
+            sacl_len = (self.offset_dacl - self.offset_sacl) if self.offset_dacl > 0 else (self.end - self.start - self.offset_sacl)
+            self.sacl = cur.decode_bytes(sacl_len)
+        if self.offset_dacl != 0:
+            self.dacl = NT_ACL(self)
+            self.dacl.decode(cur)
+        if self.end is not None:
+            self.other = cur.decode_bytes(self.end - cur)
+
+    def _encode(self, cur):
+        cur.encode_uint8le(self.revision)
+        cur.encode_uint8le(0)
+        cur.encode_uint16le(self.control)
+        cur.encode_uint32le(self.offset_owner)
+        cur.encode_uint32le(self.offset_group)
+        cur.encode_uint32le(self.offset_sacl)
+        cur.encode_uint32le(self.offset_dacl)
```

Comment:
you need to create `hole`s for these values and then fill them in later after you have encoded the previous values into the buffer and _know_ what the offset should be. See the `_encode` routine for `SessionSetupRequest`:

https://github.com/emc-isilon/pike/blob/master/pike/smb2.py#L725-L735

Reply:
accept and fixed